ion-schema/src/model/constraints/any_of.rs

```diff
@@ -0,0 +1,174 @@
+        // TODO: Replace the vec with a HashBag or come up with a less hacky way to sort the type arguments.
+        type_arguments.sort_by_cached_key(|item| format!("{item:?}"));
+
+        Self { type_arguments }
+    }
+
+    pub fn type_arguments(&self) -> impl Iterator<Item = &TypeArgument> {
+        self.type_arguments.iter()
+    }
+}
+
+impl<V: IslVersion> TypeDefinitionBuilder<V> {
+    pub fn any_of<T: Into<VersionedTypeArgumentList<V>>>(self, type_arguments: T) -> Self {
+        let type_arguments = Versioned::into_inner(type_arguments.into());
+        self.with_constraint(AnyOf::new(type_arguments).into())
+    }
+}
+
+impl ValidateInternal for AnyOf {
+    fn validate_internal<'top: 'call, 'call, R>(
+        &'top self,
+        value: &IonSchemaElement<'top>,
+        ctx: &ValidationContext,
+        recorder: &'call mut R,
+    ) -> ControlFlow<()>
+    where
+        R: ViolationRecorder<'top>,
+    {
+        todo!("Implement this once the ValidationContext allows looking up a type")
+    }
+}
+
+impl<V: IslVersion> WriteAsIsl<V> for AnyOf
+where
+    TypeArgument: WriteAsIsl<V>,
+{
+    fn write_as_isl<W: ValueWriter>(
+        &self,
+        writer: W,
+        ctx: &WriteContext<V>,
+    ) -> IonSchemaResult<()> {
+        todo!()
+    }
+}
+
+impl<V: IslVersion> ReadConstraint<V> for AnyOf {
+    fn read_constraint(ion: &Element, ctx: &LoaderContext<V>) -> IonSchemaResult<Option<Self>> {
+        todo!()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::isl::isl_type_reference::NullabilityModifier;
+    use crate::model::{IntoTypeArgument, TypeDefinition, TypeReference};
+    use crate::{ISL_1_0, ISL_2_0};
+
+    use super::*;
+
+    #[test]
+    fn test_builder() {
+        let type_ = TypeDefinitionBuilder::<ISL_1_0>::new()
+            .any_of((
+                // This is a heterogeneous "list" (really a tuple) of things that can become a type arg.
+                "foo_type",
+                TypeDefinitionBuilder::new().build(),
+                TypeReference::imported("bar.isl", "baz"),
+                "quux_type".or_null(true),
+            ))
+            .build();
+
+        assert_eq!(
+            type_.constraints().cloned().collect::<Vec<_>>(),
+            vec![AnyOf::new(vec![
+                TypeArgument::from_type_ref(
+                    NullabilityModifier::Nothing,
+                    TypeReference::from("foo_type")
+                ),
+                TypeArgument::from_type_def(
+                    NullabilityModifier::Nothing,
+                    TypeDefinition::new(vec![], vec![])
+                ),
+                TypeArgument::from_type_ref(
+                    NullabilityModifier::Nothing,
+                    TypeReference::imported("bar.isl", "baz")
+                ),
+                TypeArgument::from_type_ref(
+                    NullabilityModifier::Nullable,
+                    TypeReference::from("quux_type")
+                ),
+            ],)
```

Comment:
These explicitly-constructed `vec!`s would be good candidates for the tuple-of-`Into<_>`s trick. (Not in this PR.)

Reply:
It could be, but I'm intentionally not using it here so that if there's something wrong with the tuple-of-`Into<_>`s logic, it won't silently pass these tests. In other words, I'm intentionally constructing these the verbose way in order to make the tests more thorough.